src/plugins/workspace/public/components/workspace_creator/utils.ts

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { i18n } from '@osd/i18n';
+import {
+  WorkspacePermissionMode,
+  OptionIdToWorkspacePermissionModesMap,
+  PermissionModeId,
+} from '../../../../../core/public';
+import {
+  UserOrGroupPermissionEditingData,
+  WorkspaceFormEditingData,
+  WorkspaceFormErrors,
+  WorkspacePermissionSetting,
+  WorkspacePermissionItemType,
+  WorkspaceFormData,
+} from './types';
+
+export const isValidWorkspacePermissionSetting = (
+  setting: Partial<WorkspacePermissionSetting>
+): setting is WorkspacePermissionSetting =>
+  !!setting.modes &&
+  setting.modes.length > 0 &&
+  ((setting.type === WorkspacePermissionItemType.User && !!setting.userId) ||
+    (setting.type === WorkspacePermissionItemType.Group && !!setting.group));
+
+export const getErrorsCount = (formErrors: WorkspaceFormErrors) => {
+  let errorsCount = 0;
+  if (formErrors.name) {
+    errorsCount += 1;
+  }
+  if (formErrors.description) {
+    errorsCount += 1;
+  }
+  if (formErrors.userPermissions) {
+    errorsCount += formErrors.userPermissions.filter((permission) => !!permission).length;
+  }
+  if (formErrors.groupPermissions) {
+    errorsCount += formErrors.groupPermissions.filter((permission) => !!permission).length;
+  }
+  return errorsCount;
+};
+export const getUserAndGroupPermissions = (
+  permissions: Array<Partial<WorkspacePermissionSetting>>
+) => {
+  const userPermissions: UserOrGroupPermissionEditingData = [];
+  const groupPermissions: UserOrGroupPermissionEditingData = [];
+  for (const permission of permissions) {
+    if (permission.type === WorkspacePermissionItemType.User) {
+      userPermissions.push({ id: permission.userId, modes: permission.modes });
+    }
+    if (permission.type === WorkspacePermissionItemType.Group) {
+      groupPermissions.push({ id: permission.group, modes: permission.modes });
+    }
+  }
+  return [userPermissions, groupPermissions];
+};
+
+const getUnsavedUserOrGroupPermissionChangesCount = (
```

Comment:
I'd rename `getUnsavedPermissionsCount`

Reply:
I will simplify it in this way:

Convert the original form data: `initialFormData.permissions` which type is `UserPermissionSetting` to a data structure:

```typescript
Map<string, string>
```

The key is `${type}/${userId}` or `${type}/${groupName}`, for example:

```typescript
user key: "user/u123"
group key: "group/g123"
```
The value is a merge of modes: `modes.join('/')`, for example `read/library_read`

Then I can simply iterate `currentFormData.groupPermissions` and `currentFormData.userPermissions` to find those changed/added/deleted

Reply:
I have introduced a new type named `TypelessPermissionSetting`. Please help review, thanks!

Reply:
Thanks, I will implement it.